### DISCUSSION TITLE
MoveCopyOptions and MoveOperations in MoveToAsync(String, MoveOperations, MoveCopyOptions)

### DISCUSSION THREAD

**Discussion by Ofer-Gal** [pnp/pnpcore#884] [Q&A]
Can anyone give an example of [MoveToAsync(String, MoveOperations, MoveCopyOptions)](https://pnp.github.io/pnpcore/api/PnP.Core.Model.SharePoint.IFile.html#collapsible-PnP_Core_Model_SharePoint_IFile_MoveToAsync_System_String_PnP_Core_Model_SharePoint_MoveOperations_PnP_Core_Model_SharePoint_MoveCopyOptions_)

statement that will include Overwrite, BypassApprovePermitions and AllowBrokenThickets Enum MoveOperations and "ShouldBypassSharedLocks " MoveCopyOptions

Using + or & causes errors.
Also, where can we find more examples of files operations?

Thank you

**Comment by jansenbe** [comment]
@Ofer-Gal : Basic samples are here: https://pnp.github.io/pnpcore/using-the-sdk/files-intro.html#copying-and-moving-files

What do you mean with "Using + or & causes errors"?

**Reply by Ofer-Gal** [reply]
Thank you 
How do I specify 3 MoveOperations?  MoveOperations.Overwrite and MoveOperations.BypassApprovePermitions and MoveOperations.AllowBrokenThickets  and MoveCopyOptions= ShouldBypassSharedLocks  in the same statement?

**Reply by jansenbe** [reply]
You need to combine the flags using OR:

```csharp
await testDocument.MoveToAsync(destinationServerRelativeUrl, MoveOperations.Overwrite | MoveOperations.AllowBrokenThickets | MoveOperations.BypassApprovePermission, new MoveCopyOptions { ShouldBypassSharedLocks = true});
```

Note that BypassApprovePermission was not yet correctly defined in the code, will work as of tomorrow (see https://github.com/pnp/pnpcore/commit/a7dd4e682e894eda3dfcd70c9654c335640b1019) in the next nightly

**Reply by Ofer-Gal** [reply]
Thank you VERY Much!
I notice the item ID increments when I move with  MoveOperations.Overwrite and the history of the existing is gone.
How can I keep the ID and the history? Is there a file content replacement method?

**Reply by jansenbe** [reply]
The `ID` field can always change, it's just a unique identifier at list level. Do you want to rename a file or really move it to a different location?

**Reply by jansenbe** [reply]
For the latter there's also an async option: https://pnp.github.io/pnpcore/using-the-sdk/sites-copymovecontent.html

**Reply by Ofer-Gal** [reply]
Cool! If I set IgnoreVersionHistory = false it should keep both VersionHistories?
If destinationAbsoluteUrl has a new file name and the job is for one file only, will the target be different file name?
I am going to to try that.
Thanks again

**Reply by Ofer-Gal** [reply]
Turns out I had to get the nightly version for "context.Site.CreateCopyJobsAsync" to be recognized.
Now I am testing it to overwrite like so: 
```
string destinationAbsoluteUrl = $"{context.Uri}/DPCDocuments";
var jobUris = new List<string> { $"{context.Uri}/UserUploads/{filedata.FileName}"   };
var copyJobs = await context.Site.CreateCopyJobsAsync(jobUris.ToArray(), destinationAbsoluteUrl, new CopyMigrationOptions
{
	AllowSchemaMismatch = true,
	AllowSmallerVersionLimitOnDestination = true,
	IgnoreVersionHistory = false, //This is the reason to use it !!!!!!!! Tried true got same results.
	// Note: set IsMoveMode = true to move the file(s)
	IsMoveMode = true,
	BypassSharedLock = true,
	ExcludeChildren = true,
	NameConflictBehavior = SPMigrationNameConflictBehavior.Replace
});
```
The move is done OK but I loose the source file Version History (which why I am using this) 
Any Idea how to resolve this. My users need to get a view of older versions when files are replaced. 
I am doing the move in an Azure function because normal users should not have write access to the  "DPCDocuments" library.
The older version attempted to grant "Contribute" just for the upload time but we got about 25% 401 failures.
The PnP.JS "Add Chunked" did keep the Version History as it was before the overwrite.

Any idea will help.
Thank you much

**Reply by jansenbe** [reply]
@Ofer-Gal : can you try with setting `MoveButKeepSource = true` as well and see if that helps

**Reply by Ofer-Gal** [reply]
No it did not. :-( 
Worse, it acts like a copy. Because of our 10 years retention, we will pay dearly for storage

**Reply by jansenbe** [reply]
@Ofer-Gal : I'll see if I can repro this in my tenant, but feels like this might be a SharePoint API bug and if so, there's nothing from the PnP Core SDK side we can do to fix that.

**Reply by jansenbe** [reply]
Hi @Ofer-Gal ,

Below code correctly moves a file, including it's versions (tested with major versions), to a library in another site collection or the same site collection:

```csharp
string destinationAbsoluteUrl = $"https://bertonline.sharepoint.com/sites/pnpcoresdktest/Shared Documents";
var jobUris = new List<string> { context.Uri + "/Shared Documents/" + documentName };
var copyJob = await context.Site.CreateCopyJobsAsync(jobUris.ToArray(), destinationAbsoluteUrl, new CopyMigrationOptions
{
    AllowSchemaMismatch = true,
    AllowSmallerVersionLimitOnDestination = true,
    IgnoreVersionHistory = false,
    IsMoveMode = true,
    BypassSharedLock = true,
    ExcludeChildren = true,
    NameConflictBehavior = SPMigrationNameConflictBehavior.Replace
});

await context.Site.EnsureCopyJobHasFinishedAsync(copyJob, delay);
```

**Reply by Ofer-Gal** [reply]
You missed what I asked for :-) 
I wanted the source content moved but the destination Version History to stay. Each replace will grow the history.
In our Application, users create documents based on other documents, but the people that work on the replaced file should be able to look at versions before the replace to save some work.
Thank you for the efforts

**Reply by jansenbe** [reply]
@Ofer-Gal : doesn't that than mean just updating your destination file with the source content, that will trigger a new destination file version?

**Reply by Ofer-Gal** [reply]
How do I do that using a file in another library (same site) without adding the source to the "Preservation Hold Library"
If I get the blob and upload with overwrite, then delete the file that made the blob, a copy of that will be added to the "Preservation Hold Library" and I will pay almost double for storage. (we have 10 years retention) 
Do you know a way to exclude a library from the retention?
Thanks
